File: frontend/src/api/consumer.ts
import { ConsumerService } from "@bindings/bridge";
import type { SubscriptionClient } from "@bindings/model/models";
import type { Subscription } from "./models";
import { present } from "./client";

export const getConsumerGroups = (connID: number): Promise<Subscription[]> =>
  ConsumerService.List(connID).then(present);
export const getConsumeStats = (
  connID: number,
  group: string,
): Promise<Record<string, unknown>> =>
  ConsumerService.Stats(connID, group);

/**
 * Asks every connected consumer in a group what it holds.
 *
 * One broker round trip per client, so this is only worth paying for a group
 * somebody has opened - the group list must not call it.
 */
export const getConsumerClients = (
  connID: number,
  group: string,
): Promise<SubscriptionClient[]> =>
  ConsumerService.Clients(connID, group).then(present);

/**
 * Creates a consumer group, or rewrites an existing one.
 *
 * The broker upserts either way; the two calls exist so the form can say which
 * it meant and the caller can report it. brokerAddr empty means every master.
 *
 * consumeMode here is the broadcast *permission* the config stores, not the
 * model a client reports - the driver maps BROADCASTING onto
 * consumeBroadcastEnable. Because the whole config is rewritten, an edit must
 * send the group's current permission back or it silently clears it.
 */
export const createConsumerGroup = (
  connID: number,
  group: string,
  brokerAddr: string,
  consumeMode: string,
  maxRetry: number,
): Promise<void> =>
  ConsumerService.Create(connID, { group, brokerAddr, consumeMode, maxRetry });
export const updateConsumerGroup = (
  connID: number,
  group: string,
  brokerAddr: string,
  consumeMode: string,
  maxRetry: number,
): Promise<void> =>
  ConsumerService.Update(connID, { group, brokerAddr, consumeMode, maxRetry });

export const deleteConsumerGroup = (
  connID: number,
  group: string,
  brokerAddr: string,
): Promise<void> =>
  ConsumerService.Remove(connID, group, brokerAddr);
export const resetOffset = (
  connID: number,
  group: string,
  topic: string,
  timestamp: number,
  force: boolean,
): Promise<void> =>
  ConsumerService.ResetOffset(connID, {
    group,
    topic,
    timestamp,
    force,
  });

/**
 * Copies one group's per-queue positions onto another.
 *
 * An empty destination copies every topic the source reads. `fromOffline`
 * reads the source's stored offsets rather than its live consumers, which is
 * what a group that has already been shut down needs.
 */
export const cloneOffset = (
  connID: number,
  from: string,
  to: string,
  destination: string,
  fromOffline: boolean,
): Promise<void> =>
  ConsumerService.CloneOffset(connID, { from, to, destination, fromOffline });

/**
 * Writes one queue's committed offset for a group.
 *
 * The broker is named rather than addressed: per-queue progress rows are keyed
 * by broker name, which is the only identity a message queue carries.
 */
export const setQueueOffset = (
  connID: number,
  group: string,
  topic: string,
  broker: string,
  queueId: number,
  offset: number,
): Promise<void> =>
  ConsumerService.SetQueueOffset(connID, { group, topic, broker, queueId, offset });

File: internal/bridge/consumer.go
package bridge

import (
	"context"

	"github.com/amigoer/mq-studio/internal/driver/rocketmq"
	"github.com/amigoer/mq-studio/internal/model"
	"github.com/amigoer/mq-studio/internal/service/subscription"
)

// ConsumerService exposes subscription operations to the frontend.
//
// It speaks the canonical model. Consume mode, retry limits and the
// subscription and client lists travel in the attribute map, whose keys the
// frontend's rocketmq module reads.
type ConsumerService struct {
	service *subscription.Service
}

// ConsumerInput carries a consumer group form submission.
type ConsumerInput struct {
	Group       string `json:"group"`
	BrokerAddr  string `json:"brokerAddr"`
	ConsumeMode string `json:"consumeMode"`
	MaxRetry    int    `json:"maxRetry"`
}

func (input ConsumerInput) spec() model.SubscriptionSpec {
	return model.SubscriptionSpec{
		Ref: model.SubscriptionRef{Name: input.Group},
		Attributes: map[string]string{
			rocketmq.AttrBrokerAddr:  input.BrokerAddr,
			rocketmq.AttrConsumeMode: input.ConsumeMode,
			rocketmq.AttrMaxRetry:    itoa(input.MaxRetry),
		},
	}
}

// List returns every consumer group.
func (s *ConsumerService) List(connID int) ([]*model.Subscription, error) {
	return s.service.List(context.Background(), connID)
}

// Stats returns the per-queue consume progress of a group.
func (s *ConsumerService) Stats(connID int, group string) (map[string]interface{}, error) {
	return s.service.Stats(context.Background(), connID, model.SubscriptionRef{Name: group})
}

// Create adds a consumer group on the target broker.
func (s *ConsumerService) Create(connID int, input ConsumerInput) error {
	return s.service.Create(context.Background(), connID, input.spec())
}

// Update changes an existing consumer group.
func (s *ConsumerService) Update(connID int, input ConsumerInput) error {
	return s.service.Update(context.Background(), connID, input.spec())
}

// Remove deletes a consumer group.
func (s *ConsumerService) Remove(connID int, group string, brokerAddr string) error {
	return s.service.Remove(context.Background(), connID,
		model.SubscriptionRef{Namespace: brokerAddr, Name: group})
}

// ResetOffset moves a consumer group's read position.
func (s *ConsumerService) ResetOffset(connID int, request model.ResetOffsetRequest) error {
	return s.service.ResetOffset(context.Background(), connID, request)
}

// Clients reports what each connected consumer in a group is doing: the queues
// it holds and how fast it is getting through them.
//
// It is a separate call rather than part of Detail because it costs one round
// trip per connected client, and a group list should not pay that.
func (s *ConsumerService) Clients(connID int, group string) ([]*model.SubscriptionClient, error) {
	return s.service.Clients(context.Background(), connID, model.SubscriptionRef{Name: group})
}

// CloneOffset copies one consumer group's read position onto another, which is
// how a replacement group starts where the old one is instead of replaying
// everything it already handled.
func (s *ConsumerService) CloneOffset(connID int, request model.CloneOffsetRequest) error {
	return s.service.CloneOffset(context.Background(), connID, request)
}
